cdk/src/open-data-platform/data-plane/data-import/handler-factory.ts

```diff
@@ -1,44 +1,78 @@
 import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
-import { ConnectionPool } from '@databases/pg';
 import * as AWS from 'aws-sdk';
-import { connectToDb } from '../schema/schema.handler';
+import { AnyLengthString } from 'aws-sdk/clients/comprehendmedical';
 
 // These libraries don't have types, so they are imported in a different way.
 const { chain } = require('stream-chain');
 const { streamArray } = require('stream-json/streamers/StreamArray');
 const Batch = require('stream-json/utils/Batch');
 const Pick = require('stream-json/filters/Pick');
 
+export interface ImportRequest {
+  rowOffset?: number;
+  rowLimit?: number;
+  batchSize?: number;
+}
+
+interface ImportResult {
+  processedBatchCount: number;
+  sucessfulBatchCount: number;
+  erroredBatchCount: number;
+}
+
 const S3 = new AWS.S3();
 
 /**
  * Builds a lambda handler that parses a GeoJSON file and executes a callback.
+ *
+ * The callback takes in a set of rows and should insert them into the DB using the provided db
+ * instance.
+ *
+ * You can invoke the lambda via the web console and provide an ImportRequest to skip a set of rows
+ * or limit the import to a max number. The lambda will attempt to honor the limit, but it may
+ * exceed it due to how it proccesses rows in parallel. Example:
+ *
+ * {
+ *   "rowOffset": 0,
+ *   "rowLimit": 27000,
+ *   "batchSize": 10
+ * }
+ *
  * @param s3Params - Details for how to get the GeoJSON file.
  * @param callback - Function that operates on a single element from the file.
- * @param rowLimit - Number of rows to process, which can be used for testing.
  * @returns
  */
 export const geoJsonHandlerFactory = (
   s3Params: AWS.S3.GetObjectRequest,
-  callback: (row: any, db: ConnectionPool) => Promise<void>,
-  rowLimit: number = Infinity,
-): ((event: APIGatewayProxyEvent) => Promise<APIGatewayProxyResult>) => {
+  callback: (rows: any[], db: AWS.RDSDataService) => Promise<void>,
+): ((event: ImportRequest) => Promise<APIGatewayProxyResult>) => {
   /**
    * Reads an GeoJSON file and performs the callback on each element. This does not handle any
    * processing of the data itself, but provides a DB connection to the callback to process it.
    *
-   * TODO: track how many rows were processed and avoid re-importing the old rows.
+   * TODO: track how many batches are running in parallel and stop after a limit.
    *
    * @param db - Connection pool to the DB, made available to the callback.
    * @returns
    */
-  const readGeoJsonFile = (db: ConnectionPool): Promise<number> =>
+  const readGeoJsonFile = (
```

Comment:
a bit confused about the structure of this - what does wrapping this in the const geoJsonHandlerFactory do and why put this in a variable instead of function signature?

Reply:
I guess only the `handler` needs to be wrapped in the `geoJsonHandlerFactory`, while the other functions could receive `s3Params` and `callback` through their arguments. Let me see if that makes this more readable and addresses your question.

As for why declare it as variable instead of function: just a style preference. But I'm happy to switch to functions instead.